src/main/java/io/cdap/plugin/gcp/publisher/source/PubSubDirectDStream.java

```diff
@@ -0,0 +1,268 @@
+  private final boolean autoAcknowledge;
+  private final transient SerializableFunction<PubSubMessage, T> mappingFn;
+  private StreamingContext streamingContext;
+  private String currentSnapshotName;
+  private boolean takeSnapshot;
+  private String existingSnapshot;
+
+  public PubSubDirectDStream(io.cdap.cdap.etl.api.streaming.StreamingContext context, PubSubSubscriberConfig config,
+                             long readDuration, boolean autoAcknowledge,
+                             SerializableFunction<PubSubMessage, T> mappingFn) {
+    super(context.getSparkStreamingContext().ssc(), scala.reflect.ClassTag$.MODULE$.apply(PubSubMessage.class));
+    this.streamingContext = context.getSparkStreamingContext().ssc();
+    this.config = config;
+    this.readDuration = readDuration;
+    this.context = context;
+    this.autoAcknowledge = autoAcknowledge;
+    this.mappingFn = mappingFn;
+    this.credentials = createCredentials(config.getServiceAccount(), config.isServiceAccountFilePath());
+  }
+
+  @Override
+  public Option<RDD<T>> compute(Time validTime) {
+    LOG.debug("Computing RDD for time {}.", validTime);
+    PubSubRDD pubSubRDD = new PubSubRDD(streamingContext.sparkContext(), readDuration, config, autoAcknowledge,
+                                        credentials);
+    final SerializableFunction<PubSubMessage, T> localMappingFn = mappingFn;
+    // SerializableFunction1<PubSubMessage, T> pubSubMessageTFunction1 = ;
+    RDD<T> mapped = pubSubRDD.map(localMappingFn, scala.reflect.ClassTag$.MODULE$.apply(PubSubMessage.class));
+    return Option.apply(mapped);
+  }
+
+  @Override
+  public void start() {
+    if (config.getTopic() != null) {
+      createSubscriptionIfNotPresent();
+    }
+
+    try {
+      subscriptionAdminClient = buildSubscriptionAdminClient(credentials);
+    } catch (IOException e) {
+      throw new RuntimeException("SubscriptionAdminClient creation failed.", e);
+    }
+
+    // If state with snapshot is present, seek to it.
+    this.existingSnapshot = fetchSnapShot(config.getSubscription(), context);
+    if (this.existingSnapshot != null) {
+      seekSnapshot(config.getProject(), config.getSubscription(), existingSnapshot);
+      // Can use the same snapshot till end of the batch.
+      takeSnapshot = false;
+    } else {
+      takeSnapshot = true;
+    }
+  }
+
+  @Override
+  public void stop() {
+    if (subscriptionAdminClient == null) {
+      return;
+    }
+
+    subscriptionAdminClient.close();
+  }
+
+  @Override
+  public void onBatchStarted(io.cdap.cdap.etl.api.streaming.StreamingContext streamingContext) {
+    LOG.debug("Starting a batch.");
+    if (!takeSnapshot) {
+      this.currentSnapshotName = existingSnapshot;
+      return;
+    }
+
+    String subscription = config.getSubscription();
+    this.currentSnapshotName = createAndSave(config.getProject(), subscription,
+                                             subscription + "-snapshot-" + UUID.randomUUID(), context);
```

Comment:
let's say CDAP is down so the saveState() call keeps failing. That means we'll create a new snapshot each time this method is called, will that be a problem? Will we eventually hit some max snapshot quota or do older ones automatically get deleted?

It think it would be good to separate the calls instead of having a single createAndSave method to make it more clear that there are two calls happening, both of which can potentially fail.

Reply:
Though snapshot is taken for subscription, it is tied only to topic. Any subscription on the topic could use this snapshot to seek. No strong reason to use subscription in name other than identifying them easily in the console as you mentioned. I have refactored a bit to avoid confusion between the use of various classes for snapshot. 

Separated the calls to saveState and snapshot creation. Also included a delete snapshot when state saving fails.